File: algorithms/C/Course Schedule III/Course Schedule III.js

var scheduleCourse = function(courses) {
  courses.sort((a,b)=>a[1]-b[1]);
  let pq = new PQ(), currentDay = 0;

  for(let [duration,lastDay] of courses){
    if(duration > lastDay) continue;    
 
    if(currentDay + duration <= lastDay){
      pq.offer(duration);
      currentDay += duration;
    }
    
    else if(duration < pq.peek() && pq.size())
      {
        currentDay += duration-pq.poll();
        pq.offer(duration);
      }  
  }
  return pq.size();   
};

File: algorithms/C/Course Schedule III/Course Schedule III.java
class Solution {
    public int scheduleCourse(int[][] courses) {
        Arrays.sort(courses, (a,b)->(a[1]-b[1]));
        PriorityQueue<Integer> pq = new PriorityQueue<>((a,b)->(b - a));
        int time = 0;
        for(int[] course: courses){
            time += course[0];
            pq.add(course[0]);
            if(time > course[1])
                time -= pq.poll();
        }
        return pq.size();
    }
}
